### PR TITLE
Windows icon handlers

### PR DESCRIPTION
**Pull request by Volham22**
# What has changed ?

This PR  bring icon overlays on windows to parsec.

We have two overlays:
- A check icon to show the user that its files are synchronized
- A refresh icon to show the user that its files are being synchronized

These icons are built in two dll files that are used by `explorer.exe` to know if it should add an icon or not (is the file in a parsec workspace? Is it synchronized?)

- [X] Does this PR affect the User ?

### CODE REVIEW

File: windows-icon-handler/install_dll.bat [hunk 1]
```diff
@@ -0,0 +1,2 @@
+regsvr32 /n /i:user .\x64\Debug\check-icon-handler\check-icon-handler.dll
+regsvr32 /n /i:user .\x64\Debug\refresh-icon-handler\refresh-icon-handler.dll
```
Comment (FirelightFlagboy) [inline windows-icon-handler/install_dll.bat:2]: missing newline

**Review by FirelightFlagboy — COMMENTED** [review]
I see some missing newline, everyone as `editorconfig` setup ?

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -588,3 +588,24 @@ jobs:
         if: steps.web-change.outputs.run == 'true'
         run: npm run web:release
         working-directory: oxidation/client
+
+  ##############################################################################
+  #                            🏁 Windows icon overlays                       #
+  ##############################################################################
+
+  windows-icon-overlay:
```
Comment (FirelightFlagboy) [inline .github/workflows/ci.yml:596]: Do we really want to include the build in the CI :thinking:

Comment (Volham22) [reply .github/workflows/ci.yml:596]: That's what @Max-7 asked.

File: windows-icon-handler/install_dll.bat (hunk 1 above)
Comment (FirelightFlagboy) [reply windows-icon-handler/install_dll.bat:2]: did you have `editorconfig` installed ?

Comment (Volham22) [reply windows-icon-handler/install_dll.bat]: > did you have `editorconfig` installed ?

Yes

Comment (FirelightFlagboy) [reply windows-icon-handler/install_dll.bat]: That strange because vscode would have been correctly configured :thinking:

Comment (Volham22) [reply windows-icon-handler/install_dll.bat]: I forgot to install it on windows that's why my first commits has these issues.

 >Yes, I just discovered that vscode does not add trailing newline by default :( I've configured it now. I'll add missing trainling > newline in a next commit

I just installed `editorconfig` and everything went back to normal 😄

**Review by FirelightFlagboy — CHANGES_REQUESTED** [review]
files are missing license header

File: .github/workflows/ci.yml (hunk 2 above)
Comment (Max-7) [reply .github/workflows/ci.yml:596]: Just for the packaging, not for every PR.

**Review by FirelightFlagboy — COMMENTED** [review]
Headers are missing on:

- `.idl`
- `.def`
- `.rgs`
- `.bat`

File: .github/workflows/package-ci.yml [hunk 3]
```diff
@@ -246,6 +246,16 @@ jobs:
...
 
+      - uses: microsoft/setup-msbuild@v1.1
+
+      - name: Build Check icon overlay COM object
+        run: msbuild -maxCpuCount -property:Configuration=Release windows-icon-handler\windows-icon-handler\windows-icon-handler.vcxproj
+        shell: cmd
+
+      - name: Build Refresh icon overlay COM object
+        run: msbuild -maxCpuCount -property:Configuration=Release windows-icon-handler\refresh-icon-handler\refresh-icon-handler.vcxproj
+        shell: cmd
```
Comment (FirelightFlagboy) [inline .github/workflows/package-ci.yml:257]: The artifacts are added to `Prepare artifact` tasks ?

File: .github/workflows/package-ci.yml [hunk 4]
```diff
@@ -246,6 +246,16 @@ jobs:
           name: ${{ runner.os }}-${{ runner.arch }}-wheel
           path: dist
 
+      - uses: microsoft/setup-msbuild@v1.1
```
Comment (FirelightFlagboy) [inline .github/workflows/package-ci.yml:249]: Use commit sha instead of a tag

File: packaging/windows/README.md [hunk 5]
```diff
@@ -4,6 +4,16 @@ Inspired by [Deluge NSIS installer](https://github.com/deluge-torrent/deluge/blo
...
 ## 1 - Build the application
 
+Build shell extensions that are needed to display icon overlays in windows explorer.
+Make sure you have the windows SDK installed on your local machine (if not install it using the
+Visual Studio installer). You can use `Visual Studio` to build these two dlls __in `Release` mode__
+or invoke `msbuild` from the command line (you have to manually add `msbuild` to your `PATH`).
+
+```shell
+$ msbuild -maxCpuCount -property:Configuration=Release .\windows-icon-handler\check-icon-handler\check-icon-handler.vcxproj
+$ msbuild -maxCpuCount -property:Configuration=Release .\windows-icon-handler\refresh-icon-handler\refresh-icon-handler.vcxproj
```
Comment (FirelightFlagboy) [inline packaging/windows/README.md:14]: Since you don't show the output of the commands, you can remove `$`. that way users can copy the command directly

```suggestion
msbuild -maxCpuCount -property:Configuration=Release .\windows-icon-handler\check-icon-handler\check-icon-handler.vcxproj
msbuild -maxCpuCount -property:Configuration=Release .\windows-icon-handler\refresh-icon-handler\refresh-icon-handler.vcxproj
```

File: windows-icon-handler/check-icon-handler/check-icon-handler.rc [hunk 6]
```diff
@@ -0,0 +1,143 @@
+// Microsoft Visual C++ generated resource script.
```
Comment (FirelightFlagboy) [inline windows-icon-handler/check-icon-handler/check-icon-handler.rc:1]: Missing header

File: windows-icon-handler/refresh-icon-handler/refreshiconhandler_i.h [hunk 7]
```diff
@@ -0,0 +1,240 @@
+
+
```
Comment (FirelightFlagboy) [inline windows-icon-handler/refresh-icon-handler/refreshiconhandler_i.h:2]: Missing header

**Review by FirelightFlagboy — CHANGES_REQUESTED** [review]

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -588,3 +588,23 @@ jobs:
...
+    runs-on: windows-latest
+    name: 🏁 Windows icon overlays
+    steps:
+      - uses: actions/checkout@93ea575cb5d8a053eaa0ac8fa3b40d7e05a33cc8 # pin v3.1.0
+
+      - uses: microsoft/setup-msbuild@v1.1
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply .github/workflows/ci.yml]: @Volham22 the issue is still present

File: windows-icon-handler/refresh-icon-handler/refreshiconhandler_i.h (hunk 7 above)
Comment (Volham22) [reply windows-icon-handler/refresh-icon-handler/refreshiconhandler_i.h]: This file is auto generated and shouldn't be part of our code base. I removed it

File: windows-icon-handler/check-icon-handler/check-icon-handler.rc (hunk 6 above)
Comment (Volham22) [reply windows-icon-handler/check-icon-handler/check-icon-handler.rc:1]: This one is a bit annoying :'( Visual studio overwrite our header each time I rebuild the solution.

File: .github/workflows/ci.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (Volham22) [reply .github/workflows/ci.yml]: I've changed it in `package-ci.yml` is it good now?

File: windows-icon-handler/refresh-icon-handler/resource.h [hunk 9]
```diff
@@ -0,0 +1,20 @@
+//{{NO_DEPENDENCIES}}
```
Comment (FirelightFlagboy) [inline windows-icon-handler/refresh-icon-handler/resource.h:1]: Header ?

File: windows-icon-handler/refresh-icon-handler/refresh-icon-handler.rgs [hunk 10]
```diff
@@ -0,0 +1,3 @@
+HKCR
```
Comment (FirelightFlagboy) [inline windows-icon-handler/refresh-icon-handler/refresh-icon-handler.rgs:1]: Header ?

File: windows-icon-handler/check-icon-handler/check-icon-handler.rc (hunk 6 above)
Comment (FirelightFlagboy) [reply windows-icon-handler/check-icon-handler/check-icon-handler.rc:1]: `pre-commit` hook ?

Comment (FirelightFlagboy) [reply windows-icon-handler/check-icon-handler/check-icon-handler.rc:1]: Does the other `rc` files are auto generated ?

File: windows-icon-handler/refresh-icon-handler/resource.h (hunk 9 above)
Comment (Volham22) [reply windows-icon-handler/refresh-icon-handler/resource.h:1]: This file is auto generated too. I'm wondering if we need to commit this file too. What do you think @Max-7 ?

File: windows-icon-handler/refresh-icon-handler/refresh-icon-handler.rgs (hunk 10 above)
Comment (Volham22) [reply windows-icon-handler/refresh-icon-handler/refresh-icon-handler.rgs:1]: I tried to put a header comment at the beginning of this file but is causes issues.

TL;DR this file is bundled as a ressource into `refresh-icon-handler.dll` and it's used to update windows registry when the dll is being registered by the `regsvr32` command. Putting a comment at the beginning of this file made the dll registration fail. In fact there's no comments support in this type of file as explained [here](https://stackoverflow.com/questions/11775671/what-is-the-syntax-of-a-comment-in-a-rgs-file).

**Comment by Volham22** [comment]
I have issue with Visual Studio's solution and project files `.sln` and `vcxproj` because of line feed formats (CRLF default on windows).

File: .github/workflows/package-ci.yml [hunk 11]
```diff
@@ -285,8 +285,6 @@ jobs:
           Copy-Item -Path build\uninstall_files.nsh -Destination dist
           Copy-Item -Recurse -Path build\parsec-* -Destination dist
           Copy-Item -Path build\winfsp-* -Destination dist
-          Copy-Item -Path windows-icon-handler\refresh-icon-handler\Release\refresh-icon-handler.dll -Destination dist
-          Copy-Item -Path windows-icon-handler\check-icon-handler\Release\check-icon-handler.dll -Destination dist
```
Comment (FirelightFlagboy) [inline .github/workflows/package-ci.yml]: They're removed because of the `move` we do in `windows/freeze_program` ?

Comment (Volham22) [reply .github/workflows/package-ci.yml]: yes, it's done [here](https://github.com/Scille/parsec-cloud/pull/3764/files#diff-28d5b94a76fa468e5d7c2de1fa4b8a33928f75084c3f473e45f8d76effcbbdd9R191)

Comment (FirelightFlagboy) [reply .github/workflows/package-ci.yml]: :+1:

File: .github/workflows/ci.yml (hunk 8 above)
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply .github/workflows/ci.yml]: Yes all good :+1:

File: windows-icon-handler/check-icon-handler/check-icon-handler.rc (hunk 6 above)
Comment (Volham22) [reply windows-icon-handler/check-icon-handler/check-icon-handler.rc]: Only the boilerplate code is generated. This file need to be kept it give the product name, company name etc ... I've added licence header

File: .cspell/cspell.config.yml [hunk 12]
```diff
@@ -48,36 +48,72 @@ words:
...
+  - dbproj
   - dksjdskdjs
-  - SPIHB
-  - EFGH
-  - JKLM
-  - NPQR
+  - ebug
+  - elease
+  - eleases
+  - esult
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply .cspell/cspell.config.yml:95]: Why there's both `Capitalize` and `all lower` `Release` `Debug` folder ?

File: windows-icon-handler/check-icon-handler/CheckIconHandler.cpp [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+    return S_OK;
+}
+
+HRESULT __stdcall CCheckIconHandler::GetOverlayInfo(LPWSTR pwszIconFile, int cchMax, int* pIndex, DWORD* pdwFlags)
+{
+    // TODO: set standard icon path or load it once in memory?
+    const LPWSTR path = L"C:\\Users\\Corentin\\Documents\\icon_handler\\vs\\check.ico";
+    //const LPWSTR path = L"E:\\projects\\icon_handler\\check.ico";
+
+    // TODO: check if this file has its `__parsecinfo__`, parse its json content, set the appropriate icon if needed
```
Comment (Max-7) [inline windows-icon-handler/check-icon-handler/CheckIconHandler.cpp:52]: Wrong todo: the icon is always the same (and should be included as a resource in the DLL and loaded from there). `IsMemberOf` should be updated to check for the `.__parsec_entry_info__` file.

File: .cspell/cspell.config.yml (hunk 12 above)
[the remark this answers is not in the document]
Comment (Volham22) [reply .cspell/cspell.config.yml:95]: I don't know that's the default .gitignore for Visual Studio provided by microsoft https://github.com/github/gitignore/blob/main/VisualStudio.gitignore

[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply .cspell/cspell.config.yml:95]: Add `# cspell: disable` at the top of the .gitignore that would be simpler I think

File: windows-icon-handler/check-icon-handler/CheckIconHandler.cpp (hunk 13 above)
Comment (Volham22) [reply windows-icon-handler/check-icon-handler/CheckIconHandler.cpp]: Done.

File: packaging/windows/README.md [hunk 14]
```diff
@@ -4,10 +4,20 @@ Inspired by [Deluge NSIS installer](https://github.com/deluge-torrent/deluge/blo
 
 ## 1 - Build the application
 
+Build shell extensions that are needed to display icon overlays in windows explorer.
+Make sure you have the windows SDK installed on your local machine (if not install it using the
+Visual Studio installer). You can use `Visual Studio` to build these two dlls __in `Release` mode__
+or invoke `msbuild` from the command line (you have to manually add `msbuild` to your `PATH`).
+
+```shell
+msbuild -maxCpuCount -property:Configuration=Release .\windows-icon-handler\check-icon-handler\check-icon-handler.vcxproj
```
Comment (Max-7) [inline packaging/windows/README.md:13]: Could build the sln instead of each individual vcxproj.

File: windows-icon-handler/check-icon-handler/CheckIconHandler.cpp [hunk 15]
```diff
@@ -0,0 +1,84 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
```
Comment (Max-7) [inline windows-icon-handler/check-icon-handler/CheckIconHandler.cpp:1]: Not the right header. This is the backend header, the whole `windows-icon-handler` folder should have the parsec core header.

File: windows-icon-handler/install_dll.bat [hunk 16]
```diff
@@ -0,0 +1,3 @@
+:: Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
+
```
Comment (Max-7) [inline windows-icon-handler/install_dll.bat:2]: Is this file used?

**Review by Max-7 — CHANGES_REQUESTED** [review]

File: packaging/windows/README.md (hunk 14 above)
Comment (Volham22) [reply packaging/windows/README.md:13]: I did this in https://github.com/Scille/parsec-cloud/pull/3764/commits/44963686fa2605d2612d273172b6ab691a3b1787 I'll update the `README` guide.

File: windows-icon-handler/install_dll.bat (hunk 16 above)
Comment (Volham22) [reply windows-icon-handler/install_dll.bat]: No, I removed it

File: windows-icon-handler/common/include/parsec.h [hunk 17]
```diff
@@ -0,0 +1,144 @@
+// Parsec Cloud (https://parsec.cloud) Copyright (c) AGPL-3.0 2016-present Scille SAS
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:1]: Missing include guards (ifndef or pragma once)

File: windows-icon-handler/common/include/parsec.h [hunk 18]
```diff
@@ -0,0 +1,144 @@
...
+        log_impl(ofs, x, args...);
+        ofs.flush();
+    }
+
+#define TRACE(...) ::parsec::log(__FILE__, __LINE__, __VA_ARGS__)
+#else
+#define TRACE(...) /* nothing */
+#endif
+
+    inline SyncState is_member_file(const std::wstring_view& pwsz_path)
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:76]: Not an appropriate name for its return value. If the function is named `is_member_file`, I'm expected a boolean. Rename it to something like `get_sync_state()` with `Synced`, `NotSynced`, `NotAMemberFile` or something like it.

File: windows-icon-handler/common/include/parsec.h [hunk 19]
```diff
@@ -0,0 +1,144 @@
...
+        if (drive_number == -1 || drive_number < 7 || PathIsRootW(pwsz_path.data()))
+        {
+            TRACE("Drive number is", drive_number);
+            return SyncState::NotSet;
+        }
+
+        // Parsec write a registry key to set the icon for its workspaces.
+        // We check if the key is present for the current drive.
+        auto reg_key = std::make_unique<wchar_t[]>(MAX_PATH * 32);
+        swprintf(
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:97]: Use a sstream instead.

File: windows-icon-handler/common/include/parsec.h [hunk 20]
```diff
@@ -0,0 +1,144 @@
...
+        if (RegOpenKeyEx(HKEY_CURRENT_USER, reg_key.get(), 0, KEY_READ, &key) != ERROR_SUCCESS)
+        {
+            TRACE("Call to `RegOpenKeyEx` failed, return `NotSet`");
+            return SyncState::NotSet;
+        }
+
+        auto entry_infos = std::ifstream(std::format(L"{}.__parsec_entry_info__", pwsz_path));
+
+        // Can't open file for some reason, stop there
+        if (!entry_infos.good())
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:115]: Add curly brackets.

File: windows-icon-handler/common/include/parsec.h [hunk 21]
```diff
@@ -0,0 +1,144 @@
...
+            return SyncState::NotSet;
+
+        std::stringstream buffer;
+        buffer << entry_infos.rdbuf();
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:119]: Why use this buffer? Can't you do like in the json lib example? https://github.com/nlohmann/json#read-json-from-a-file

File: windows-icon-handler/common/include/parsec.h [hunk 22]
```diff
@@ -0,0 +1,144 @@
...
+            TRACE(
+                "Json parsing OK `is_sync` value",
+                json_object["need_sync"].get<bool>(),
+                "return value is",
+                json_object["need_sync"].get<bool>() ? SyncState::Refresh : SyncState::Synced
+            );
+
+            return json_object["need_sync"].get<bool>() ? SyncState::Refresh : SyncState::Synced;
+        }
+#ifdef _DEBUG
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:134]: Why the ifdef?

File: windows-icon-handler/common/include/parsec.h [hunk 23]
```diff
@@ -0,0 +1,144 @@
...
+
+namespace parsec
+{
+    enum class SyncState {
+        Synced,
+        Refresh,
+        NotSet
+    };
+
+#ifdef _DEBUG
```
Comment (Max-7) [inline windows-icon-handler/common/include/parsec.h:24]: I think we're better off removing the debug part. It probably won't be used ever again, makes the code harder to read, will cause problems if we compile in debug by mistake, and contains hardcoded absolute paths.

File: windows-icon-handler/refresh-icon-handler/RefreshIconHandler.cpp [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    ofs.flush();
+}
+
+HRESULT __stdcall CRefreshIconHandler::IsMemberOf(LPCWSTR pwszPath, DWORD dwAttrib)
+{
+    auto state = parsec::is_member_file(pwszPath);
+    log_result(state, pwszPath);
+    if (state == parsec::SyncState::Refresh)
```
Comment (Max-7) [inline windows-icon-handler/refresh-icon-handler/RefreshIconHandler.cpp:40]: Add curly brackets.

File: windows-icon-handler/refresh-icon-handler/RefreshIconHandler.cpp [hunk 25]
```diff
@@ -0,0 +1,60 @@
+// RefreshIconHandler.cpp : Implementation of CRefreshIconHandler
+
+#include "pch.h"
+#include "RefreshIconHandler.h"
+
+#include "parsec.h"
+
+
+// CRefreshIconHandler
+static void log_result(parsec::SyncState state, std::wstring path)
```
Comment (Max-7) [inline windows-icon-handler/refresh-icon-handler/RefreshIconHandler.cpp:10]: To remove.

**Review by Max-7 — CHANGES_REQUESTED** [review]

File: windows-icon-handler/common/include/parsec.h (hunk 22 above)
Comment (Volham22) [reply windows-icon-handler/common/include/parsec.h]: There're not needed now because we remove the debug part.

**Review by Max-7 — APPROVED** [review]

**Comment by Volham22** [comment]
> It's expected to include external/include/json/json.hpp ? can't this be installed by a dependencies manager ? or a script ?
This `c++` part isn't big enough to justify the use of a C++ package manager. I think we can add a build script that download the released header on github and launch the build.

**Comment by Volham22** [comment]
> > It's expected to include external/include/json/json.hpp ? can't this be installed by a dependencies manager ? or a script ?
> 
> ~~This `c++` part isn't big enough to justify the use of a C++ package manager. I think we can add a build script that download the released header on github and launch the build instead.~~

I discussed with @Max-7 and it turns out we don't want to maintain a build script for this extension given we are not supposed to modify this extension.

**Review by Max-7 — APPROVED** [review]

**Comment by FirelightFlagboy** [comment]
@Volham22 silly me, Can you make a quick PR to add the missing news fragment for this issue ?
